src/Core/Rendering/RenderController.cpp

```diff
@@ -708,12 +709,51 @@ namespace MxEngine
         blurInputOutput->Bind(textureId++);
         applyShader->SetUniform("inputTex", input->GetBoundId());
         applyShader->SetUniform("SSGITex", blurInputOutput->GetBoundId());
-
+         
         this->RenderToTexture(output, applyShader);
 
         std::swap(input, output);
     }
+    void RenderController::ApplyDepthOfFieldEffect(CameraUnit& camera, TextureHandle& io, TextureHandle& tmp0, TextureHandle& tmp1)
+    {
+        if (camera.Effects->GetFocusDistance() == 0.f)
+            return;
+        MAKE_SCOPE_PROFILER("RenderController::ApplyDepthOfFieldEffect()");
+
+        io->GenerateMipmaps();
+        auto cocShader = this->Pipeline.Environment.Shaders["COC"_id];
+        cocShader->Bind(); 
+        cocShader->IgnoreNonExistingUniform("camera.viewProjMatrix");
+        cocShader->IgnoreNonExistingUniform("normalTex"); 
+        cocShader->IgnoreNonExistingUniform("albedoTex");
+        cocShader->IgnoreNonExistingUniform("materialTex"); 
+        Texture::TextureBindId textureId = 0;
+        this->BindGBuffer(camera, *cocShader, textureId);
+        io->Bind(textureId++); 
+        cocShader->SetUniform("cameraOutput", io->GetBoundId());
+        cocShader->SetUniform("focusRange", camera.Effects->GetFocusRange());        
+        cocShader->SetUniform("focusDistance", camera.Effects->GetFocusDistance());
+        this->BindCameraInformation(camera, *cocShader);
+        this->RenderToTextureNoClear(tmp0, cocShader);
+
+        auto bokehShader = this->Pipeline.Environment.Shaders["Bokeh"_id];
+        bokehShader->Bind();
+        bokehShader->SetUniform("bokehRadius", camera.Effects->GetBokehRadius());
+        
+        io->Bind(0);
+        tmp0->Bind(1);
+        this->RenderToTextureNoClear(tmp1, bokehShader); 
+        
+        this->ApplyGaussianBlur(tmp1, tmp0, 1);
```

Comment:
oh, if you are using gaussian blur, it will drop alpha channel information (see gaussian_blur_fragment.glsl)
You can either modify the gauss shader to preserve it without blurring, or store the alpha value (COC in your case) in a separate texture whichs not blurred
With my comment below about combining two shaders, I guess its better to modify gauss shader to preserve alpha